iotileemulate/iotile/emulate/internal/rpc_queue.py

```diff
@@ -177,6 +188,9 @@ def start(self):
                 self._rpc_queue.task_done()
             except AsynchronousRPCResponse:
                 self._queue_async_rpc(address, rpc_id, response)
+            except BusyRPCResponse:
```

Comment:
This extra `BusyRPCResponse` block shouldn't be necessary since the generic `except` block below has the same functionality.  `response.capture_exception()` is just a synonym for `response.set_exception` for use inside of an except clause.

Reply:
The thing is with _raise_exception:

    def _raise_exception(self):
        if self._exception is None:
            raise InternalError("No exception stored in call to _raise_exception")

        exc_type, value, traceback = self._exception

        if value is not None and isinstance(exc_type, Exception):
            raise TypeError("instance exception may not have a separate value")

        if value is not None:
            exc = exc_type(value)
        else:
            exc = exc_type

        if exc.__traceback__ is not traceback:
            raise exc.with_traceback(traceback)

        raise exc

That I don't understand intention behind the line, to me it resembles copy constructor but it's not the case in python

    exc = exc_type(value)

And because of it I receive an exception

    (Pdb) Traceback (most recent call last):
      File "/opt/poduino/coretools/iotileemulate/.tox/py36/lib/python3.6/site-packages/iotile/core/hw/transport/virtualadapter.py", line 392, in send_rpc_async
        response = dev.call_rpc(address, rpc_id, bytes(payload))
      File "/opt/poduino/coretools/iotileemulate/.tox/py36/lib/python3.6/site-packages/iotile/emulate/virtual/emulated_device.py", line 193, in call_rpc
        return self.emulator.call_rpc_external(address, rpc_id, payload)
      File "/opt/poduino/coretools/iotileemulate/.tox/py36/lib/python3.6/site-packages/iotile/emulate/internal/emulation_loop.py", line 259, in call_rpc_external
        return response.wait(timeout)
      File "/opt/poduino/coretools/iotileemulate/.tox/py36/lib/python3.6/site-packages/iotile/emulate/internal/response.py", line 129, in wait
        self._raise_exception()
      File "/opt/poduino/coretools/iotileemulate/.tox/py36/lib/python3.6/site-packages/iotile/emulate/internal/response.py", line 62, in _raise_exception
        exc = exc_type(value)
    TypeError: __init__() takes 1 positional argument but 2 were given
    *** END EXCEPTION ***


May you suggest if  I missed something in BusyResponseException implementation?

Reply:
I just ran into this bug as well and fixed it in #722 so after that is merged `response.capture_exception` should work.

Reply:
I see, now it makes sense